src/modules/flow/twitter/twitter.c

```diff
@@ -0,0 +1,617 @@
+    SOL_NULL_CHECK_GOTO(blob, err_blob);
+
+    r = sol_message_digest_feed(digest, blob, true);
+    SOL_INT_CHECK_GOTO(r, < 0, err_feed);
+
+    sol_blob_unref(blob);
+    free(escaped_status);
+    free(escaped_params);
+    free(params);
+
+    return 0;
+
+err_feed:
+    sol_blob_unref(blob);
+err_blob:
+    free(signature);
+err_signature:
+    free(escaped_params);
+err_escape_params:
+    free(params);
+err_params:
+    sol_message_digest_del(digest);
+err_digest:
+    free(escaped_status);
+err_escape_status:
+    free(cb_data->status);
+err_status:
+    free(cb_data->key);
+err_key:
+    free(cb_data->timestamp);
+err_timestamp:
+    free(cb_data->nonce);
+err:
+    free(cb_data);
+    return r;
+}
+
+static int
+token_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    int r;
+    uint16_t idx;
+    const char *val;
+    size_t token_len, token_secret_len;
+    struct sol_buffer buf;
+    struct sol_vector tokens;
+    struct sol_str_slice *slice;
+    struct twitter_data *mdata = data;
+
+    free(mdata->token);
+    mdata->token = NULL;
+    free(mdata->token_secret);
+    mdata->token_secret = NULL;
+
+    r = sol_flow_packet_get_string(packet, &val);
+    SOL_INT_CHECK(r, < 0, r);
+
+    token_len = sizeof("oauth_token=") - 1;
+    token_secret_len = sizeof("oauth_token_secret=") - 1;
+
+    tokens = sol_util_str_split(
+        sol_str_slice_from_str(val), "&", 0);
+    SOL_VECTOR_FOREACH_IDX (&tokens, slice, idx) {
+        struct sol_str_slice token;
+        if (streqn(slice->data, "oauth_token_secret=", token_secret_len) &&
+            !mdata->token_secret) {
+            token = SOL_STR_SLICE_STR(slice->data + token_secret_len,
+                slice->len - token_secret_len);
+            r = sol_http_encode_slice(&buf, token);
+            SOL_INT_CHECK_GOTO(r, < 0, err);
+            mdata->token_secret = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
+            sol_buffer_fini(&buf);
+            SOL_NULL_CHECK_GOTO(mdata->token_secret, err);
+        } else if (streqn(slice->data, "oauth_token=", token_len) &&
+            !mdata->token) {
+            token = SOL_STR_SLICE_STR(slice->data + token_len, slice->len - token_len);
+            r = sol_http_encode_slice(&buf, token);
+            SOL_INT_CHECK_GOTO(r, < 0, err);
+            mdata->token = sol_str_slice_to_string(sol_buffer_get_slice(&buf));
+            sol_buffer_fini(&buf);
+            SOL_NULL_CHECK_GOTO(mdata->token, err);
+        }
+    }
+
+    if (!mdata->token || !mdata->token_secret) {
+        r = -EINVAL;
+        goto err;
+    }
+
+    sol_vector_clear(&tokens);
+    return 0;
+
+err:
+    free(mdata->token);
+    free(mdata->token_secret);
+    sol_vector_clear(&tokens);
+    return r;
+}
+
+static int
+timeline_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
```

Comment:
this is almost a replica of the other function above, please refactor to share these common bits.